src/controllers/bulk/bulkenumerator.cpp

```diff
@@ -2,60 +2,73 @@
-    for (int i = 0; bulk_supported[i].vendor_id; ++i) {
-        if ((bulk_supported[i].vendor_id == desc->idVendor) &&
-            (bulk_supported[i].product_id == desc->idProduct)) {
-            return true;
-        }
-    }
-    return false;
+static bool is_interesting(const libusb_device_descriptor& desc) {
+    auto vendorId = desc.idVendor;
+    auto productId = desc.idProduct;
+    return std::any_of(std::cbegin(bulk_supported),
+            std::cend(bulk_supported),
+            [=](const auto& dev) {
+                return dev.vendor_id == vendorId && dev.product_id == productId;
+            });
 }
 
 QList<Controller*> BulkEnumerator::queryDevices() {
     qDebug() << "Scanning USB Bulk devices:";
-    libusb_device **list;
-    ssize_t cnt = libusb_get_device_list(m_context, &list);
-    ssize_t i = 0;
-    int err = 0;
-
-    for (i = 0; i < cnt; i++) {
-        libusb_device *device = list[i];
+    libusb_device** deviceListPointers; // contiguous array of pointers to libusb_device
+    ssize_t cnt = libusb_get_device_list(m_context.get(), &deviceListPointers);
+    VERIFY_OR_DEBUG_ASSERT(cnt >= 0) {
+        // this assumes that if libusb_get_device_list returned an error, no memory was allocated
+        // and listPtr is still uninitialized. Unfortunately the libusb docs don't specify the state
+        // of the arguments in the case of errors.
+        qWarning() << "Error retrieving devicelist: " << libusb_error_name(cnt);
+    }
+    auto pDeviceList = std::unique_ptr<libusb_device*[],
+            decltype([](libusb_device** list) {
+                libusb_free_device_list(list, true);
+            })>(deviceListPointers);
+    m_devices.clear();
+    for (libusb_device* pDevice : std::span(pDeviceList.get(), cnt)) {
         struct libusb_device_descriptor desc;
 
-        libusb_get_device_descriptor(device, &desc);
-        if (is_interesting(&desc)) {
-            struct libusb_device_handle* handle = nullptr;
-            err = libusb_open(device, &handle);
-            if (err) {
-                qWarning() << "Error opening a device:" << libusb_error_name(err);
-                continue;
-            }
-
-            BulkController* currentDevice =
-                    new BulkController(m_context, handle, &desc);
-            m_devices.push_back(currentDevice);
+        libusb_get_device_descriptor(pDevice, &desc);
+        if (!is_interesting(desc)) {
+            continue;
         }
+        struct libusb_device_handle* handle = nullptr;
+        int err = libusb_open(pDevice, &handle);
+        if (err) {
+            qWarning() << "Error opening a device:" << libusb_error_name(err);
+            continue;
+        }
+        m_devices.push_back(std::make_unique<BulkController>(m_context.get(), handle, &desc));
+    }
+    QList<Controller*> devices;
+    devices.reserve(m_devices.size());
+    for (const auto& pController : std::as_const(m_devices)) {
```

Comment:
Can we get rid of this loop by returning a const refrences to m_devices? Before it was not an issue because of QTs implicit sharing.

Reply:
Well that would blow up the scope because I would need to refactor the the API to use something other than `QList`. That is a long-term goal which I wanted to avoid for now because the container is usually very small (less then two dozen elements). I don't think this stopgap is an issue for now.